### PR TITLE
Tezos Recovery, Solana Tokens & Missing Requirements

### PR DESCRIPTION
**Pull request by yanlevi**
Added support for recovering XTZ base asset.

### CODE REVIEW

File: utils/sol_helpers.py [hunk 1]
```diff
@@ -17,19 +22,38 @@
...
+    return response
+
+
+def withdraw_token(priv, transfer_params: TransferCheckedParams) -> RPCResponse:
+    solana_client = Client(URL)
+    blockhash = get_blockhash()
+    txn = Transaction().add(transfer_checked(transfer_params))
+    txn.recent_blockhash = blockhash
+    txn.fee_payer = transfer_params.owner
+    txn.instructions[0].keys[1].is_writable = True
```
Comment (arielt1) [inline utils/sol_helpers.py:48]: Please add a comment explaining the reason for setting it writeable

File: utils/sol_helpers.py [hunk 2]
```diff
@@ -17,19 +22,38 @@
...
+    signature = eddsa_sign.eddsa_sign(priv, txn.serialize_message())
+    final_sig = Signature(signature)
```
Comment (arielt1) [inline utils/sol_helpers.py:50]: Why not do it in a single step?

File: utils/sol_helpers.py [hunk 3]
```diff
@@ -44,29 +68,64 @@ def get_blockhash():
...
+        amount: Human readable amount. Will be manipulated according to decimals.
+        token_address: The address of the token. ("Smart Contract")
+        decimals: Can be found under the token description in any blockchain explorer.
+
+    Returns:
+
+    """
+    path = f'{BIP_44_CONSTANT}/{SOL_ASSET_NUM}/{account}/{CHANGE}/{ADDR_INDEX}'
+    priv, pub = eddsa_sign.eddsa_derive(xpriv, path)
+    true_amount = int(amount * (10 ** decimals))
```
Comment (arielt1) [inline utils/sol_helpers.py:121]: Maybe call it "scaled_amount"?

File: utils/xtz_helpers.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+BIP_44_CONSTANT = 44
+XTZ_ASSET_NUM = 1729
+CHANGE = 0
+ADDR_INDEX = 0
+RPC = "https://rpc.tzbeta.net/"
+
+
+def tb(l):
+    return b''.join(map(lambda x: x.to_bytes(1, 'big'), l))
```
Comment (arielt1) [inline utils/xtz_helpers.py:18]: Please use more verbose naming for function name and variables/parameters

File: utils/xtz_helpers.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+
+base58_encodings = [
+    #    Encoded   |               Decoded             |
+    # prefix | len | prefix                      | len | Data type
```
Comment (arielt1) [inline utils/xtz_helpers.py:23]: Maybe use named tuple?

File: utils/xtz_helpers.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    'delegation': 3
+}
+
+
+def base58_encode(v: bytes, prfx: bytes) -> bytes:
+    try:
+        encoding = next(
+            encoding
+            for encoding in base58_encodings
+            if len(v) == encoding[3] and prfx == encoding[0]
```
Comment (arielt1) [inline utils/xtz_helpers.py:81]: named tuple might be more readable here

File: utils/xtz_helpers.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    return b58encode_check(encoding[2] + v)
+
+
+def base58_decode(v: bytes) -> bytes:
+    try:
+        prefix_len = next(
+            len(encoding[2])
+            for encoding in base58_encodings
+            if len(v) == encoding[1] and v.startswith(encoding[0])
```
Comment (arielt1) [inline utils/xtz_helpers.py:94]: same here

File: utils/sol_helpers.py (hunk 2 above)
Comment (yanlevi) [reply utils/sol_helpers.py]: Done

File: utils/sol_helpers.py (hunk 3 above)
Comment (yanlevi) [reply utils/sol_helpers.py]: Done

File: utils/xtz_helpers.py (hunk 4 above)
Comment (yanlevi) [reply utils/xtz_helpers.py]: Done

File: utils/xtz_helpers.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        )
+    except StopIteration:
+        raise ValueError('Invalid encoding, prefix or length mismatch.')
+
+    return b58decode_check(v)[prefix_len:]
+
+
+def get_xtz_pub_key(public_key: bytes) -> bytes:
+    return b58encode_check(b'\r\x0f%\xd9' + public_key)
```
Comment (arielt1) [inline utils/xtz_helpers.py:103]: What is `b'\r\x0f%\xd9'`? Could it be defined as some constant?

File: utils/xtz_helpers.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    """
+
+    :param derived_prv:
+    :param derived_pub:
+    :param xtz_amount:
+    :param xtz_destination:
+    :return:
+    """
+    xtz_address = get_xtz_address(derived_pub)
+    pytezos - pytezos.using(shell=RPC, key=xtz_address)
```
Comment (arielt1) [inline utils/xtz_helpers.py:142]: I think you meant `=`

File: utils/xtz_helpers.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    filled_transaction_opg = transaction_opg.autofill()
+    # *** 1. Watermark message ***
+    validation_pass = validation_passes[filled_transaction_opg.contents[0]['kind']]
+    if any(map(lambda x: validation_passes[x['kind']] != validation_pass, filled_transaction_opg.contents)):
+        raise ValueError('Mixed validation passes')
+
+    print(validation_pass, "is the validation pass")
+    if validation_pass == 0:
+        chain_watermark = bytes.fromhex(filled_transaction_opg.shell.chains.main.watermark())
+        watermark = b'\x02' + chain_watermark
```
Comment (arielt1) [inline utils/xtz_helpers.py:153]: What is `b'\x02'`? Can it be moved to a constant/global?

File: utils/xtz_helpers.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    else:
+        watermark = b'\x03'
```
Comment (arielt1) [inline utils/xtz_helpers.py:155]: Same here

**Review by arielt1 — APPROVED** [review]

**Review by cblokh — APPROVED** [review]

**Pull request merged**